beacon-chain/state/state-native/getters_validator.go

```diff
@@ -145,6 +145,22 @@ func (b *BeaconState) PubkeyAtIndex(idx primitives.ValidatorIndex) [fieldparams.
 	return bytesutil.ToBytes48(b.validators[idx].PublicKey)
 }
 
+// PublicKeys builds a list of all validator public keys, with each key's index aligned to its validator index.
+func (b *BeaconState) PublicKeys() [][fieldparams.BLSPubkeyLength]byte {
```

Comment:
Don't we have this cached? this is a large slice, it's probably better to grab the cache first, updating it here on calls to this functions with the new pubkeys from the state, and return the cache.

Reply:
For backfill service, this is only called once at start-up, validator public keys re cached in `verify.go`'s `verifier`'s `keys`. With that said, I think your suggestion is still valid

Reply:
We only run this once at startup, for the checkpoint state we're backfilling from. I don't know if it will be in cache at that point.